enhance grouping for context menu options

sikhote opened the pull request:
### Description

- Lowers the minimum children count to 2, allowing for smaller groups within context menus
- Adds a separator between groups
- Add documentation and some variable renaming for clarity

### Screenshots:

Before:
![Screenshot 2023-05-15 at 2 04 54 PM](https://github.com/opensearch-project/OpenSearch-Dashboards/assets/20453492/18b1a884-29b5-461c-afe1-e5543d4aa071)


After:
![Screenshot 2023-05-15 at 2 03 26 PM](https://github.com/opensearch-project/OpenSearch-Dashboards/assets/20453492/06ca0796-a603-4161-834d-0bce7488ad98)



### Issues Resolved

fixes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3171

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

sikhote commented:
@ashwin-pc @joshuarrrr this is ready for final review.

ashwin-pc commented on src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:
```diff
@@ -83,6 +83,12 @@ interface Props {
   SavedObjectFinder: React.ComponentType<any>;
   stateTransfer?: EmbeddableStateTransfer;
   hideHeader?: boolean;
+  // By default, embeddable.destroy() is called when this component unmounts.
+  // To prevent this default behavior, set this prop to true.
+  isDestroyPrevented?: boolean;
+  // Toggle off the border and shadow applied around the embeddable.
+  // By default, the embeddable will have a shadow and border around it.
+  isBorderless?: boolean;
```
Hey @sikhote, why are these two changes necessary?

sikhote replied:
In the process of checking creating an example of this in the `Examples` section, I discovered there is an existing `EmbeddableRenderer` that does the job. These changes are now removed.

ashwin-pc commented on src/plugins/ui_actions/public/actions/action_internal.ts:
```diff
@@ -59,7 +59,7 @@ export class ActionInternal<A extends ActionDefinition = ActionDefinition>
   }
 
-  public getDisplayName(context: Context<A>): string {
+  public getDisplayName(context: Context<A>): JSX.Element | string {
```
Can we rename this since Display name is misleading here? `getDisplayElement` perhaps?

sikhote replied:
Agreed that name would better, but I'd recommend against it at this point, because it would mean a lot of downstream changes as well (such as other plugins not in this repository).

ashwin-pc reviewed (commented):
@sikhote can you provide steps in the PR so that i can validate the change? Also i dont see any example in here on how you plan on using the jsx property of `getDisplayName`. Im wondering if we can scope this down since allowing any arbitrary JSX element to appear in place of a name can have unintended consequences.

sikhote commented:
> @sikhote can you provide steps in the PR so that i can validate the change? Also i dont see any example in here on how you plan on using the jsx property of `getDisplayName`. Im wondering if we can scope this down since allowing any arbitrary JSX element to appear in place of a name can have unintended consequences.

The UI actions & context menu changes can be seen when locally at [http://localhost:5601/app/uiActionsExplorer](http://localhost:5601/app/uiActionsExplorer).

Regarding our usage of `getDisplayName`, we looking to implement the designs within this issue: https://github.com/opensearch-project/alerting-dashboards-plugin/issues/483

Specifically, this design:
https://user-images.githubusercontent.com/105245012/224512024-a1095c23-c107-4a07-9ee1-50711c211550.png

In order to accomplish creating the `Documentation` link with an icon, we need to use a React component. This is supported by the [Eui documentation](https://eui.elastic.co/v34.6.0/#/navigation/context-menu).

@lezzago if you think we can scope this down, it would reduce the changes here, but we would miss out on that functionality.

sikhote commented on src/plugins/embeddable/public/lib/embeddables/embeddable_factory.ts:
```diff
@@ -81,7 +81,7 @@ export interface EmbeddableFactory<
    * Returns a display name for this type of embeddable. Used in "Create new... " options
```
If it helps, I can change this comment to inform users that a React element is also supported here.

ashwin-pc replied:
Do we need this change? This is for the Create new button for embeddables right?

sikhote commented on src/plugins/ui_actions/public/actions/action.ts:
```diff
@@ -92,7 +92,7 @@ export interface Action<Context extends BaseContext = {}, T = ActionType>
    * Returns a title to be displayed to the user.
```
If it helps, I can change this comment to inform users that a React element is also supported here.

joshuarrrr replied:
Yeah, that'd be helpful.

ashwin-pc commented on src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.tsx:
```diff
@@ -157,35 +157,51 @@ export async function buildContextMenuForActions({
       for (let i = 0; i < action.grouping.length; i++) {
         const group = action.grouping[i];
-        currentPanel = group.id;
-        if (!panels[currentPanel]) {
+        const groupId = group.id;
+
+        if (!panels[groupId]) {
           const name = group.getDisplayName ? group.getDisplayName(context) : group.id;
-          panels[currentPanel] = {
-            id: currentPanel,
+
+          // Create panel for group
+          panels[groupId] = {
+            id: groupId,
             title: name,
             items: [],
             _level: i,
             _icon: group.getIconType ? group.getIconType(context) : 'empty',
           };
-          if (parentPanel) {
-            panels[parentPanel].items!.push({
+
+          // If there are multiple groups and this is not the first group,
+          // then add an item to the parent panel relating to this group
+          if (parentPanelId) {
+            panels[parentPanelId].items!.push({
               name,
-              panel: currentPanel,
+              panel: groupId,
               icon: group.getIconType ? group.getIconType(context) : 'empty',
               _order: group.order || 0,
               _title: group.getDisplayName ? group.getDisplayName(context) : '',
             });
           }
         }
-        parentPanel = currentPanel;
+
+        // Save the current panel, because this will be used for adding items
+        // to it from later groups in the array
+        parentPanelId = groupId;
```
Just to make sure, nothing functionally chnages here right? You are just renaming currentPannel to groupId

sikhote replied:
Correct, renaming and adding comments for future users.

ashwin-pc reviewed (commented):
The PR looks good to me, just a few comments:

1. The change for embeddables seems unnecessary, but let me know if im wrong there
2. Can you add a test for the grouping change too? We dont want a future change breaking that behaviour for you right?
3. Can you update the PR description to reflect the correctly reflect the current set of changes, the embeddable panel change can be removed.

sikhote commented:
@ashwin-pc @lezzago I've completed the following
- Updated description
- Added some additional code comments
- Expanded the tests to be more specific for testing certain use-cases

ashwin-pc reviewed (approved):
This looks great! thanks for the changes.

ashwin-pc commented:
@sikhote can you resolve the changelog conflicts?

joshuarrrr commented on src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.test.ts:
```diff
@@ -248,3 +256,195 @@ test('hides items behind in "More" submenu if there are more than 4 actions', as
+  `);
+});
+
+test('grouping with only two actions', async () => {
```
thanks for all the tests!

joshuarrrr reviewed (approved):
Nice, thanks @sikhote !

sikhote commented:
Thanks for the help ;)
